src/com/amazon/corretto/crypto/provider/AesCfbSpi.java

```diff
@@ -0,0 +1,333 @@
+      return parameters;
+    } catch (final InvalidParameterSpecException | NoSuchAlgorithmException e) {
+      throw new RuntimeCryptoException("Unexpected error", e);
+    }
+  }
+
+  @Override
+  protected void engineInit(final int opmode, final Key key, final SecureRandom random)
+      throws InvalidKeyException {
+    try {
+      byte[] iv = new byte[IV_SIZE_IN_BYTES];
+      random.nextBytes(iv);
+      engineInit(opmode, key, new IvParameterSpec(iv), random);
+    } catch (InvalidAlgorithmParameterException e) {
+      throw new InvalidKeyException("Failed to initialize with random IV", e);
+    }
+  }
+
+  @Override
+  protected void engineInit(
+      final int opmode,
+      final Key key,
+      final AlgorithmParameterSpec params,
+      final SecureRandom random)
+      throws InvalidKeyException, InvalidAlgorithmParameterException {
+    if (!(params instanceof IvParameterSpec)) {
+      throw new InvalidAlgorithmParameterException("Params must be an instance of IvParameterSpec");
+    }
+    final IvParameterSpec ivParameterSpec = (IvParameterSpec) params;
+    final byte[] ivBytes = ivParameterSpec.getIV();
+    init(opmode, key, ivBytes);
+  }
+
+  @Override
+  protected void engineInit(
+      final int opmode, final Key key, AlgorithmParameters params, final SecureRandom random)
+      throws InvalidKeyException, InvalidAlgorithmParameterException {
+    if (params == null) {
+      throw new InvalidAlgorithmParameterException("Params must not be null");
+    }
+    try {
+      engineInit(opmode, key, params.getParameterSpec(IvParameterSpec.class), random);
+    } catch (final InvalidParameterSpecException e) {
+      throw new InvalidAlgorithmParameterException(e);
+    }
+  }
+
+  private void init(final int opmode, final Key key, final byte[] ivBytes)
+      throws InvalidKeyException, InvalidAlgorithmParameterException {
+    if (ivBytes.length != IV_SIZE_IN_BYTES) {
+      throw new InvalidAlgorithmParameterException(
+          "Provided IV must be of length " + IV_SIZE_IN_BYTES);
+    }
+
+    if (!"RAW".equalsIgnoreCase(key.getFormat())) {
+      throw new InvalidKeyException("Key's format must be RAW");
+    }
+    final byte[] keyBytes = key.getEncoded();
+    if (keyBytes == null) {
+      throw new InvalidKeyException("Key must support encoding");
+    }
+    if (keyBytes.length != KEY_LEN_AES128 && keyBytes.length != KEY_LEN_AES256) {
+      throw new InvalidKeyException(
+          "Key length must be " + KEY_LEN_AES128 + " or " + KEY_LEN_AES256);
+    }
+
+    switch (opmode) {
+      case Cipher.ENCRYPT_MODE:
+        this.opMode = ENC_MODE;
+        break;
+      case Cipher.DECRYPT_MODE:
+        this.opMode = DEC_MODE;
+        break;
+      default:
+        throw new InvalidAlgorithmParameterException("Invalid opmode: " + opmode);
+    }
+    this.keyLen = keyBytes.length;
+    this.key = keyBytes.clone();
+    this.ivParamSpec = new IvParameterSpec(ivBytes); // TODO [childw] enforce IV constraints
```

Comment:
Has this TODO been fixed?

Reply:
ah, thanks for the catch. that's an artifact of [refactoring](https://github.com/corretto/amazon-corretto-crypto-provider/pull/458/commits/219889ff4c2e063d572d95dd185def1104a3ca92#diff-0fca1c846fc2c12ed2a4161a86bfaaa862e98e768f923cd80a74defd2a8853bdR169) the validation logic down into the common `private void init(...)`. removed.